src/Workspaces/Core/Portable/Workspace/Workspace_Events.cs

```diff
@@ -59,22 +78,40 @@ protected Task RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind kind, Solutio
             projectId = documentId.ProjectId;
         }
 
-        var ev = GetEventHandlers<WorkspaceChangeEventArgs>(WorkspaceChangeEventName);
+        var ev = GetEventHandlers<WorkspaceChangeEventArgs>(WorkspaceChangedImmediateEventName);
+        WorkspaceChangeEventArgs args = null;
+        if (ev.HasHandlers)
+        {
+            args = new WorkspaceChangeEventArgs(kind, oldSolution, newSolution, projectId, documentId);
+            RaiseEventForHandlers(ev, args, isImmediate: true);
+        }
+
+        ev = GetEventHandlers<WorkspaceChangeEventArgs>(WorkspaceChangeEventName);
         if (ev.HasHandlers)
         {
             return this.ScheduleTask(() =>
             {
-                using (Logger.LogBlock(FunctionId.Workspace_Events, (s, p, d, k) => $"{s.Id} - {p} - {d} {kind.ToString()}", newSolution, projectId, documentId, kind, CancellationToken.None))
-                {
-                    var args = new WorkspaceChangeEventArgs(kind, oldSolution, newSolution, projectId, documentId);
-                    ev.RaiseEvent(static (handler, arg) => handler(arg.self, arg.args), (self: this, args));
-                }
+                args ??= new WorkspaceChangeEventArgs(kind, oldSolution, newSolution, projectId, documentId);
+                RaiseEventForHandlers(ev, args, isImmediate: false);
             }, WorkspaceChangeEventName);
         }
         else
         {
             return Task.CompletedTask;
         }
+
+        static void RaiseEventForHandlers(
+            EventMap.EventHandlerSet<EventHandler<WorkspaceChangeEventArgs>> handlers,
+            WorkspaceChangeEventArgs args,
+            bool isImmediate)
+        {
+            var functionId = isImmediate ? FunctionId.Workspace_EventsImmediate : FunctionId.Workspace_Events;
+
+            using (Logger.LogBlock(functionId, (s, p, d, k) => $"{s.Id} - {p} - {d} {args.Kind.ToString()}", args.NewSolution, args.ProjectId, args.DocumentId, args.Kind, CancellationToken.None))
+            {
+                handlers.RaiseEvent(static (handler, args) => handler(args.NewSolution.Workspace, args), args);
```

Comment:
I def have concerns about exceptions. Can we try/catch with a fatalerror reporter?

Reply:
From the other thread:

Any concern with me just changing RaiseEvent from using FatalError.ReportAndPropagate to instead use FatalError.ReportAndCatch (also, would it make sense to have the try/catch inside the foreach in that method?)

Reply:
yes.  please ReportAndCatch. We really do not our eventing to ever get into a failed state (or corrupt the WS).